### PR TITLE
New hook: useSafeTimeout

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Moves the logic within `withSafeTimeout` to the new hook `useSafeTimeout`.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/compose/src/hooks/use-safe-timeout/index.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * WordPress dependencies
+ */
+import { useRef, useEffect } from '@wordpress/element';
+
+export default function useSafeTimeout() {
+	const timeouts = useRef( [] );
+
+	function clearTimeout( id ) {
```
Comment (ellatrix) [inline packages/compose/src/hooks/use-safe-timeout/index.js]: Should we use `useCallback` for these?

**Review by youknowriad — COMMENTED** [review]
To be fair,  I'm not sure whether we really need this hook. I feel this is just a Higher-order that was needed because of how class components work. I feel with hooks, we don't really need it.

**Comment by ellatrix** [comment]
@youknowriad Could you elaborate? Why won't we need it?

**Comment by youknowriad** [comment]
I feel when you use timeouts with hooks, you'll always do something like that;

```
useEffect(() =>  {
  const tid = setTimeout(dosomething);
  return () => clearTimeout(tid);
});
```

The timeout and its clearance are located in the same code, the chance to forget about this and the added code here is very small while in the case of a class Component or components without hooks, it's quite complex to achieve: you need refactor to a class component and use separate methods to trigger and clear the timeouts. All these are fixed with hooks.

**Comment by ellatrix** [comment]
Ah, you're right. I was thinking of a use case in RichText where `setTimeout` wouldn't be in `useEffect`, but this seems very rare.

**Comment by aduth** [comment]
> Ah, you're right. I was thinking of a use case in RichText where `setTimeout` wouldn't be in `useEffect`, but this seems very rare.

There's now another example of this (timers in event handler) at #19881, though I'd agree that it's still not yet (to me) at a point to need a higher-level solution, and if anything its absence could serve as a deterrent to an anti-pattern.

**Pull request closed without merging**